### PR TITLE
First pass for machine deployment controller

### PR DESCRIPTION
**Pull request by k4leung4**
**What this PR does / why we need it**:
This is the first pass for the machine deployment controller.

Copies/borrows heavily from deployment controller.

Things not implemented in this PR:
- deployment progress
- orphan adoption

Caveat
- rolling update does not work due to machineset status not implemented

@kubernetes/kube-deploy-reviewers

### CODE REVIEW

File: pkg/controller/machinedeployment/sync.go [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	apirand "k8s.io/apimachinery/pkg/util/rand"
+
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	dutil "sigs.k8s.io/cluster-api/pkg/controller/machinedeployment/util"
+)
+
+// syncStatusOnly only updates Deployments Status and doesn't take any mutating actions.
+func (dc *MachineDeploymentControllerImpl) syncStatusOnly(d *v1alpha1.MachineDeployment, msList []*v1alpha1.MachineSet, machineMap map[types.UID]*v1alpha1.MachineList) error {
```
Comment (dgoodwin) [inline pkg/controller/machinedeployment/sync.go]: Possible unused function here.

File: pkg/controller/machinedeployment/util/util.go [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+			// We deterministically choose the oldest new MachineSet.
+			return msList[i]
+		}
+	}
+	// new MachineSet does not exist.
+	return nil
+}
+
+// FindOldMachineSets returns the old machine sets targeted by the given Deployment, with the given slice of MSes.
+// Note that the first set of old machine sets doesn't include the ones with no machines, and the second set of old machine sets include all old machine sets.
```
Comment (dgoodwin) [inline pkg/controller/machinedeployment/util/util.go]: Suggest rewording that this returns a list of all old machine sets with replicas scaled to 0, and a list includes all old machine sets regardless of replicas.

File: pkg/controller/machinedeployment/sync.go [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+		}
+		return nil
+	}
+
+	// There are old machine sets with machines and the new machine set is not saturated.
+	// We need to proportionally scale all machine sets (new and old) in case of a
+	// rolling deployment.
+	if dutil.IsRollingUpdate(deployment) {
+		allMSs := dutil.FilterActiveMachineSets(append(oldMSs, newMS))
+		allMSsReplicas := dutil.GetReplicaCountForMachineSets(allMSs)
```
Comment (dgoodwin) [inline pkg/controller/machinedeployment/sync.go]: Suggest a rename to "totalMSReplicas".

File: pkg/controller/machinedeployment/sync.go [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+
+	newDeployment := d
+	newDeployment.Status = newStatus
+	_, err := dc.machineClient.ClusterV1alpha1().MachineDeployments(newDeployment.Namespace).UpdateStatus(newDeployment)
+	return err
+}
+
+// calculateStatus calculates the latest status for the provided deployment by looking into the provided machine sets.
+func calculateStatus(allMSs []*v1alpha1.MachineSet, newMS *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) v1alpha1.MachineDeploymentStatus {
+	availableReplicas := dutil.GetAvailableReplicaCountForMachineSets(allMSs)
```
Comment (dgoodwin) [inline pkg/controller/machinedeployment/sync.go]: PR mentioned MachineSet Status not implemented yet, is this still accurate and the code is just prepared for when AvailableReplicas is properly set?

File: pkg/controller/machinedeployment/util/util.go [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+	return apiequality.Semantic.DeepEqual(t1Copy, t2Copy)
+}
+
+// FindNewMachineSet returns the new MS this given deployment targets (the one with the same machine template).
+func FindNewMachineSet(deployment *v1alpha1.MachineDeployment, msList []*v1alpha1.MachineSet) *v1alpha1.MachineSet {
+	sort.Sort(MachineSetsByCreationTimestamp(msList))
+	for i := range msList {
+		if EqualIgnoreHash(&msList[i].Spec.Template, &deployment.Spec.Template) {
+			// In rare cases, such as after cluster upgrades, Deployment may end up with
+			// having more than one new MachineSets that have the same template as its template,
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/util/util.go]: done

File: pkg/controller/machinedeployment/util/util.go [hunk 6]
```diff
@@ -0,0 +1,691 @@
...
+			// see https://github.com/kubernetes/kubernetes/issues/40415
+			// We deterministically choose the oldest new MachineSet.
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/util/util.go]: done

File: pkg/controller/machinedeployment/util/util.go [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+	t1Copy := template1.DeepCopy()
+	t2Copy := template2.DeepCopy()
+	// Remove hash labels from template.Labels before comparing
+	delete(t1Copy.Labels, v1alpha1.DefaultMachineDeploymentUniqueLabelKey)
+	delete(t2Copy.Labels, v1alpha1.DefaultMachineDeploymentUniqueLabelKey)
+	return apiequality.Semantic.DeepEqual(t1Copy, t2Copy)
+}
+
+// FindNewMachineSet returns the new MS this given deployment targets (the one with the same machine template).
+func FindNewMachineSet(deployment *v1alpha1.MachineDeployment, msList []*v1alpha1.MachineSet) *v1alpha1.MachineSet {
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/util/util.go]: Not sure if "current" is really better. IMHO, it all comes down to terminology.
"new" refers to the new machine set that we are upgrading to, and "old" refers to the old machine set that we are upgrading from.
In a sense, "current" runs into the same issue that both the machineset that we are upgrading from and upgrading to are current. 
I am all for better naming, as I too had difficulty following the deployment controller code, where this code is based off of. Though I am not sure if renaming it to "current" helps much.

[the remark this answers is not in the document]
Comment (dgoodwin) [reply pkg/controller/machinedeployment/util/util.go]: That sounds ok, new is very pervasive and consistent here so this sounds ok to me now.

File: pkg/controller/machinedeployment/sync.go (hunk 4 above)
Comment (k4leung4) [reply pkg/controller/machinedeployment/sync.go]: This is still accurate and you are correct.
The MachineSet Status has the fields for AvailableReplicas, but the MachineSet controller currently does not populate it.
The method pulls value out of the MachineSet Status, but since it isn't populated, it always returns 0, which is the default.

File: pkg/controller/machinedeployment/sync.go (hunk 1 above)
Comment (k4leung4) [reply pkg/controller/machinedeployment/sync.go]: you are correct, removing unused code.

File: pkg/controller/machinedeployment/sync.go (hunk 3 above)
Comment (k4leung4) [reply pkg/controller/machinedeployment/sync.go]: done

File: pkg/controller/machinedeployment/util/util.go (hunk 2 above)
Comment (k4leung4) [reply pkg/controller/machinedeployment/util/util.go]: done.

File: pkg/controller/machinedeployment/controller.go [hunk 8]
```diff
@@ -15,41 +14,328 @@ See the License for the specific language governing permissions and
...
+			continue
+		}
+		deployments = append(deployments, d)
+	}
+
+	return deployments
+}
+
+// updateMachineSet figures out what deployment(s) manage a MachineSet when the MachineSet
+// is updated and wake them up. If the anything of the MachineSets have changed, we need to
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/controller.go]: That is correct.
updated godoc

File: pkg/controller/machinedeployment/util/util.go [hunk 9]
```diff
@@ -0,0 +1,691 @@
...
+	sort.Sort(MachineSetsByCreationTimestamp(msList))
+	for i := range msList {
+		if EqualIgnoreHash(&msList[i].Spec.Template, &deployment.Spec.Template) {
+			// In rare cases, such as after cluster upgrades, Deployment may end up with
+			// having more than one new MachineSets that have the same template as its template,
+			// see https://github.com/kubernetes/kubernetes/issues/40415
+			// We deterministically choose the oldest new MachineSet.
+			return msList[i]
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/util/util.go]: I don't think it is to say choosing the oldest matching MachineSet is always the correct choice. I too believe that there are scenarios where it is correct to choose the newest one.
The choice to choose the oldest one here is for it to be deterministic, instead of a random behavior that is dependent on the ordering of the list.

File: pkg/controller/machinedeployment/controller.go [hunk 10]
```diff
@@ -15,41 +14,328 @@ See the License for the specific language governing permissions and
...
-// Reconcile handles enqueued messages
+func (c *MachineDeploymentControllerImpl) getMachineSetsForDeployment(d *v1alpha1.MachineDeployment) ([]*v1alpha1.MachineSet, error) {
+	// List all MachineSets to find those we own but that no longer match our
+	// selector. They will be orphaned by ClaimMachineSets().
+	msList, err := c.msLister.MachineSets(d.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: flush out machine set adoption.
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/controller.go]: For pod deployments, replicasets, the use case is that the user created a replicaset by hand, and wanted to upgrade it, and so they would create a pod deployment with matching labels, so it could adopt it and then upgrade using deployment strategies.

For machine deployments, it would be the equivalent of someone creating a machineset by hand, and wanting to upgrade it using machine deployment strategies.

File: pkg/controller/machinedeployment/controller.go [hunk 11]
```diff
@@ -15,41 +14,328 @@ See the License for the specific language governing permissions and
...
+	}
+
+	// If it has a ControllerRef, that's all that matters.
+	if controllerRef := metav1.GetControllerOf(ms); controllerRef != nil {
+		d := c.resolveControllerRef(ms.Namespace, controllerRef)
+		if d == nil {
+			return
+		}
+		glog.V(4).Infof("MachineSet %s added for deployment %v.", ms.Name, d.Name)
+		c.Reconcile(d)
```
Comment (csrwng) [inline pkg/controller/machinedeployment/controller.go]: Instead of calling Reconcile directly, you should queue the MachineDeployment. Otherwise, you could have Reconcile called simultaneously on the same MachineDeployment. (Same applies to all calls to Reconcile from event handlers)

Comment (csrwng) [reply pkg/controller/machinedeployment/controller.go]: You can get access to the queue in your `Init` function with:
```go
queue := arguments.GetSharedInformers().WorkerQueues["MachineDeployment"].Queue
```

File: pkg/controller/machinedeployment/controller.go [hunk 12]
```diff
@@ -15,41 +14,344 @@ See the License for the specific language governing permissions and
...
+	msList, err := c.msLister.MachineSets(d.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: flush out machine set adoption.
+
+	var filteredMS []*v1alpha1.MachineSet
+	for _, ms := range msList {
+		if metav1.GetControllerOf(ms) != nil && !metav1.IsControlledBy(ms, d) {
```
Comment (justinsb) [inline pkg/controller/machinedeployment/controller.go]: If a MS doesn't have a controller, we want to ignore it, I think?  So `if metav1.GetController(ms) == nil || ...`

File: pkg/controller/machinedeployment/util/util.go [hunk 13]
```diff
@@ -0,0 +1,693 @@
...
+	case 0:
+		// If there is no active machine set then we should return the newest.
+		if newMS != nil {
+			return newMS
+		}
+		return oldMSs[0]
+	case 1:
+		return allMSs[0]
+	default:
+		return nil
```
Comment (justinsb) [inline pkg/controller/machinedeployment/util/util.go]: Is this behaviour intended?  It doesn't really match the comment I think - allMSs[0] is the latest I think

File: pkg/controller/machinedeployment/controller.go (hunk 12 above)
Comment (k4leung4) [reply pkg/controller/machinedeployment/controller.go]: nice catch, was originally thinking of handling it when I added the adoption code, but it makes sense to ignore it for now.
fixed.

File: pkg/controller/machinedeployment/controller.go [hunk 14]
```diff
@@ -15,41 +14,344 @@ See the License for the specific language governing permissions and
...
-	log.Printf("Running reconcile MachineDeployment for %s\n", u.Name)
-	return nil
+	// Deep-copy otherwise we are mutating our cache.
+	d := u.DeepCopy()
+
+	everything := metav1.LabelSelector{}
+	if reflect.DeepEqual(d.Spec.Selector, &everything) {
+		if d.Status.ObservedGeneration < d.Generation {
+			d.Status.ObservedGeneration = d.Generation
+			c.machineClient.ClusterV1alpha1().MachineDeployments(d.Namespace).UpdateStatus(d)
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/controller.go]: Return on err, added logging.

File: pkg/controller/machinedeployment/controller.go [hunk 15]
```diff
@@ -15,41 +14,344 @@ See the License for the specific language governing permissions and
...
+}
+
+// resolveControllerRef returns the controller referenced by a ControllerRef,
+// or nil if the ControllerRef could not be resolved to a matching controller
+// of the correct Kind.
+func (c *MachineDeploymentControllerImpl) resolveControllerRef(namespace string, controllerRef *metav1.OwnerReference) *v1alpha1.MachineDeployment {
+	// We can't look up by UID, so look up by Name and then verify UID.
+	// Don't even try to look up by Name if it's the wrong Kind.
+	if controllerRef.Kind != controllerKind.Kind {
+		return nil
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/controller.go]: agree, added.

File: pkg/controller/machinedeployment/controller.go [hunk 16]
```diff
@@ -15,41 +14,344 @@ See the License for the specific language governing permissions and
...
+	}
+	d, err := c.mdLister.MachineDeployments(namespace).Get(controllerRef.Name)
+	if err != nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/controller.go]: do we want to retry? if the machine deployment is gone, would this loop indefinitely?

File: pkg/controller/machinedeployment/controller_test.go [hunk 17]
```diff
@@ -0,0 +1,896 @@
...
+		},
+		Spec: v1alpha1.MachineDeploymentSpec{
+			Strategy: v1alpha1.MachineDeploymentStrategy{
+				Type: common.RollingUpdateMachineDeploymentStrategyType,
+				RollingUpdate: &v1alpha1.MachineRollingUpdateDeployment{
+					MaxUnavailable: func() *intstr.IntOrString { i := intstr.FromInt(0); return &i }(),
+					MaxSurge:       func() *intstr.IntOrString { i := intstr.FromInt(0); return &i }(),
+				},
+			},
+			Replicas: func() *int32 { i := int32(replicas); return &i }(),
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/controller_test.go]: the inline thing was bugging me as well, fixed.

File: pkg/controller/machinedeployment/controller_test.go [hunk 18]
```diff
@@ -0,0 +1,896 @@
...
+}
+
+func TestSyncDeploymentCreatesMachineSet(t *testing.T) {
+	f := newFixture(t)
+
+	d := newMachineDeployment("foo", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	f.dLister = append(f.dLister, d)
+	f.objects = append(f.objects, d)
+
+	ms := newMachineSet(d, "deploymentms-4186632231", 1)
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/controller_test.go]: random name, changed it to be less cryptic.

File: pkg/controller/machinedeployment/controller_test.go [hunk 19]
```diff
@@ -0,0 +1,896 @@
...
+
+			if test.expectedUIDs != len(machineMap) {
+				t.Fatalf("got %v machine set UIDs, expected %v machine set UIDs", test.expectedUIDs, len(machineMap))
+			}
+
+			if test.numMS1Machines != len(machineMap[ms1.UID].Items) {
+				t.Fatalf("got %v machines, expected %v machines for ms1", test.numMS1Machines, len(machineMap[ms1.UID].Items))
+			}
+			if test.numMS2Machines != len(machineMap[ms2.UID].Items) {
+				t.Fatalf("got %v machines, expected %v machines for ms1", test.numMS1Machines, len(machineMap[ms1.UID].Items))
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/controller_test.go]: thanks for catching that, fixed.

File: pkg/controller/machinedeployment/rolling.go [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+	allMSs := append(oldMSs, newMS)
+
+	// Scale up, if we can.
+	scaledUp, err := dc.reconcileNewMachineSet(allMSs, newMS, d)
+	if err != nil {
+		return err
+	}
+	if scaledUp {
+		return nil
+		// Update DeploymentStatus
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/rolling.go]: Forgot about that.
Added TODO to update deployment status for deployment progress.

File: pkg/controller/machinedeployment/rolling.go [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+	if oldMachinesCount == 0 {
+		// Can't scale down further
+		return false, nil
+	}
+
+	allMachinesCount := dutil.GetReplicaCountForMachineSets(allMSs)
+	glog.V(4).Infof("New machine set %s/%s has %d available machines.", newMS.Namespace, newMS.Name, newMS.Status.AvailableReplicas)
+	maxUnavailable := dutil.MaxUnavailable(*deployment)
+
+	// Check if we can scale down. We can scale down in the following 2 cases:
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/rolling.go]: This is the same logic as the one from deployment.
Would the max unavailable setting help prevent the worst case scenario, where some number of machines needs to be healthy before proceeding?

File: pkg/controller/machinedeployment/rolling.go [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+		}
+	}
+
+	return nil
+	// Sync deployment status
+	//return dc.syncRolloutStatus(allMSs, newMS, d)
+}
+
+func (dc *MachineDeploymentControllerImpl) reconcileNewMachineSet(allMSs []*v1alpha1.MachineSet, newMS *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) (bool, error) {
+	if *(newMS.Spec.Replicas) == *(deployment.Spec.Replicas) {
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/rolling.go]: Replicas is not nil, as it is pulled from working machinesets, though to better guard it, added the helper.

File: pkg/controller/machinedeployment/rolling.go [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+	sort.Sort(dutil.MachineSetsByCreationTimestamp(oldMSs))
+
+	totalScaledDown := int32(0)
+	totalScaleDownCount := availableMachineCount - minAvailable
+	for _, targetMS := range oldMSs {
+		if totalScaledDown >= totalScaleDownCount {
+			// No further scaling required.
+			break
+		}
+		if *(targetMS.Spec.Replicas) == 0 {
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/rolling.go]: that is some confusing code, pulled out some additional variables to hopefully make it easier to follow

yes, this was copied as is from the deployment code.

File: pkg/controller/machinedeployment/sync.go [hunk 24]
```diff
@@ -0,0 +1,434 @@
...
+		}
+
+		return nil, err
+	case err != nil:
+		glog.V(4).Infof("Failed to create new machine set %q: %v", newMS.Name, err)
+		return nil, err
+	}
+
+	needsUpdate := dutil.SetDeploymentRevision(d, newRevision)
+	if !alreadyExists && d.Spec.ProgressDeadlineSeconds != nil {
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/sync.go]: no reason, removed.

File: pkg/controller/machinedeployment/sync.go [hunk 25]
```diff
@@ -0,0 +1,434 @@
...
+		// Set existing new machine set's annotation
+		annotationsUpdated := dutil.SetNewMachineSetAnnotations(d, msCopy, newRevision, true)
+
+		minReadySecondsNeedsUpdate := msCopy.Spec.MinReadySeconds != *d.Spec.MinReadySeconds
+		if annotationsUpdated || minReadySecondsNeedsUpdate {
+			msCopy.Spec.MinReadySeconds = *d.Spec.MinReadySeconds
+			return dc.machineClient.ClusterV1alpha1().MachineSets(msCopy.ObjectMeta.Namespace).Update(msCopy)
+		}
+
+		// Should use the revision in existingNewMS's annotation, since it set by before
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/sync.go]: updated the comment.

we attempt to copy over the revision annotation over to the deployment if it doesn't already have it.

File: pkg/controller/machinedeployment/util/util.go (hunk 13 above)
Comment (k4leung4) [reply pkg/controller/machinedeployment/util/util.go]: The naming was off. It really wanted one active or latest. If there are multiple machine sets that are available, it needs to start scaling down some of the older machine sets first to avoid having too many active machine sets.

renamed the method and updated the comment.

File: pkg/controller/machinedeployment/rolling_test.go [hunk 26]
```diff
@@ -0,0 +1,422 @@
...
+		readyMachinesFromOldMS int
+		readyMachinesFromNewMS int
+		scaleExpected          bool
+		expectedOldReplicas    int
+		expectedActions        int
+	}{
+		{
+			name:                   "scenario 1: 10 desired, oldMS at 10, 10 ready, 0 max unavailable => oldMS at 9, scale down by 1.",
+			deploymentReplicas:     10,
+			maxUnavailable:         intstr.FromInt(0),
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/rolling_test.go]: yes, that is an invalid configuration. this would normally have been safe guarded at a higher level.
fixed test.

File: pkg/controller/machinedeployment/rolling_test.go [hunk 27]
```diff
@@ -0,0 +1,422 @@
...
+		expectedOldReplicas int
+	}{
+		{
+			name:                "scenario 1. 10 desired, oldMS at 10, 10 ready, max unavailable 0 => oldMS at 9, scale down by 1.",
+			deploymentReplicas:  10,
+			maxUnavailable:      intstr.FromInt(0),
+			readyMachines:       10,
+			oldReplicas:         10,
+			scaleExpected:       true,
+			expectedOldReplicas: 9,
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/rolling_test.go]: updated test.

File: pkg/controller/machinedeployment/controller_test.go [hunk 28]
```diff
@@ -0,0 +1,896 @@
...
+
+func hasExpectedMachineNames(t *testing.T, mList *v1alpha1.MachineList, msName string, numMSMachines int) {
+	var names, expectedNames []string
+	for _, m := range mList.Items {
+		names = append(names, m.Name)
+	}
+	sort.Strings(names)
+
+	for i := 0; i < numMSMachines; i++ {
+		expectedNames = append(expectedNames, fmt.Sprintf("%v-machine-%v", msName, i))
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/controller_test.go]: at the moment, it will probably error out or get truncated. if it is the latter, it might get confusing.
i'll file an issue to make sure we handle the case, that is check the generated name does not exceed the max length, and that the spec in deployment/machineset have base names that does is short enough to allow a hash to be appended.

File: pkg/controller/machinedeployment/rolling.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (justinsb) [reply pkg/controller/machinedeployment/rolling.go]: Given this is the deployment model and that seems to work in practice, I think that trumps my intuition :-)  I'm sure we'll learn new things when we do this on real machines!

**Comment by justinsb** [comment]
Thanks for the changes.

LGTM - but it's so big I could easily have missed something.  As it's new code, I vote to merge and we can find things though - particularly as this is based on the (pod) deployment code.

File: pkg/controller/machinedeployment/rolling.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machinedeployment/rolling.go]: We will definitely learn more once we get people using it.
After we get more usage, and deem that we want to stay in sync with the deployment controller logic, I will be reaching out to the kubernetes team to talk about possibilities of sharing their logic to avoid the duplication here.

**Comment by k4leung4** [comment]
Thanks for taking a stab at this huge PR.
